### ISSUE TITLE
Some suggestions about Submit

### ISSUE THREAD

**Issue by YehudahA** [AngleSharp/AngleSharp#218]
Let's see this case:

Html:

``` html
<form>
    <input type="text" name="UserName" />
    <select name="City">
        <option value="0">London</option>
        <option value="1">Jerusalem</option>
    </select>
</form>
```

C# code:

``` cs
var page = await form.Submit(new Dictionary<string, string>
{
    ["User"] = "Yehudah",
    ["City"] = "1"
});
```

My problems are:
1. The code ignores silently the no-matched field.
2. It doesn't set the SELECT field.
3. In addition, I suggest to extract this logic from `Submit` function, because I want to use the same logic, but then `GetSubmission` and change some aspect of the request.

I suggest this code:

``` cs
public static Task<IDocument> Submit(this IHtmlFormElement form, IDictionary<string, string> fields)
{
    form.SetFields(fields);
    return form.Submit();
}

public static void SetFields(this IHtmlFormElement form, IDictionary<string, string> fields, bool createIfNoFaund = false)
{
    var elements = form.Elements.OfType<IHtmlInputElement>();
    var selects = form.QuerySelectorAll<IHtmlSelectElement>("select");
    form.Submit(new Dictionary<string, string>());
    foreach (var field in fields)
    {
        var element = elements.FirstOrDefault(e => e.Name == field.Key);
        var select = selects.FirstOrDefault(s => s.Name == field.Key);

        if (element != null)
        {
            element.Value = field.Value;
        }
        else if (select != null)
        {
            select.Value = field.Value;
        }
        else
        {
            if (createIfNoFaund)
            {
                var newElementHtml = $"<input type=\"hidden\" name=\"{field.Key}\" value=\"{field.Value}\" />";
                form.Insert(AdjacentPosition.BeforeEnd, newElementHtml);
            }
            else
            {
                throw new KeyNotFoundException($"field {field.Key} not found");
            }
        }
    }
}
```

**Comment by FlorianRappl** [comment]
Thanks for the suggestions. If you want you can submit a PR request - even with C# 6. Beginning in the new year I will change the complete code base to use C# 6 constructs. The suggestion will then be part of v0.9.5.

Only a couple of points that need to be included in such a PR:
- Do not run queries before they might be needed (potentially we do not need all selects + why are we limiting the form elements?)
- Same applies to LINQ queries
- Why do you submit the form before setting the fields? Is that a mistake? Should that line be in there at all?
- What about `select` elements that belong to that form, but have been placed outside of the `form` element? Usually, the `form.Elements` property should return _all_ elements, i.e., including `select`, `input`, ... types. (The elements of this list also contain input elements that have been placed outside of the corresponding `form`, however, with the `form` attribute being set properly).

Thanks again for your suggestions! :+1:

**Comment by YehudahA** [comment]
I created PR.
I modified the code in accordance with your comments.

**Comment by FlorianRappl** [comment]
Thanks for your efforts! See my comments in #219.

**Comment by FlorianRappl** [comment]
Has been merged into `devel`, see #220.
